src/test/java/ru/curs/celesta/transaction/CelestaTransactionTest.java

```diff
@@ -78,8 +89,54 @@ void handlesErrorInMethod() {
                 }));
     }
 
+    @Test
+    void handlesRollbackErrorInMethod() {
+        this.contextRunner
+                .withConfiguration(UserConfigurations.of(DummyService.class))
+                .withPropertyValues("celesta.h2.inMemory:true")
+                .run((context -> {
+                    try (Celesta celesta = context.getBean(Celesta.class)) {
+                        DummyService srv = context.getBean(DummyService.class);
+                        CallContext cc = new SystemCallContext(){
+                            @Override
+                            public void rollback(){
+                                throw new IllegalStateException("no rollback");
+                            }
+                        };
+                        assertFalse(cc.isClosed());
+                        assertEquals("no rollback",
+                                assertThrows(IllegalStateException.class,
+                                        () -> srv.exception(cc)
+                                ).getMessage());
+                        assertTrue(cc.isClosed());
+                        shutDownH2(celesta);
+                    }
+                }));
+    }
+
     private void shutDownH2(Celesta celesta) throws SQLException {
         celesta.getConnectionPool().get().createStatement().execute("SHUTDOWN");
     }
 
+    @Test
+    void handlesCallContextActivationFailure() {
+        Celesta celesta = mock(Celesta.class);
+        when(celesta.getConnectionPool()).thenThrow(new IllegalStateException("no connections"));
+        when(celesta.getProfiler()).thenReturn(mock(IProfiler.class));
+        CallContext ctx = new SystemCallContext();
+
+        ProceedingJoinPoint jp = mock(ProceedingJoinPoint.class);
+        Signature signature = mock(Signature.class);
+        when(signature.toShortString()).thenReturn("foo");
+        when(jp.getArgs()).thenReturn(new Object[]{ctx});
+        when(jp.getSignature()).thenReturn(signature);
+
+        CelestaTransactionAspect aspect = new CelestaTransactionAspect(celesta);
+        assertFalse(ctx.isClosed());

```

Comment:
`assertAll`?

Reply:
Not for this case, I think... first we assert that context is not closed, then perform a transaction, then assert that it is closed. The order of execution of asserts matters here and when one of the asserts fails, there is no reason to proceed with test and checks.